### PR TITLE
Opensearch Reject Execution Exception handling

### LINKED ISSUE #627: [BUG] Replication stops with NotSerializableExceptionWrapper exception
**What is the bug?**

Replication stops with org.opensearch.common.io.stream.NotSerializableExceptionWrapper exception

```
{
  "status" : "PAUSED",
  "reason" : """AutoPaused: [[xx-rr][18] - org.opensearch.common.io.stream.NotSerializableExceptionWrapper - "replication_exception: failed to replay changes"], [[xx-rr][62] - org.opensearch.common.io.stream.NotSerializableExceptionWrapper - "replication_exception: failed to replay changes"], """,
  "leader_alias" : "xx",
  "leader_index" : "xx",
  "follower_index" : "xxrr"
}
```

And when I try to resume it ,

```
{
  "error" : {
    "root_cause" : [
      {
        "type" : "resource_not_found_exception",
        "reason" : "Retention lease doesn't exist. Replication can't be resumed for xxx-rr"
      }
    ],
    "type" : "resource_not_found_exception",
    "reason" : "Retention lease doesn't exist. Replication can't be resumed for xxx-rr"
  },
  "status" : 404
}
```

And it also does auto pause

```
{
  "status" : "PAUSED",
  "reason" : """AutoPaused: [[xx-rr-1][23] - org.opensearch.OpenSearchException - "rejected execution of org.opensearch.common.util.concurrent.TimedRunnable@72fa7b3a on EWMATrackingOpenSearchThreadPoolExecutor[name = 562a078f676b91f136fdd28bf0f75be7/write, queue capacity = 10000, task execution EWMA = 401.1ms, org.opensearch.common.util.concurrent.EWMATrackingOpenSearchThreadPoolExecutor@44415adc[Running, pool size = 32, active threads = 32, queued tasks = 10004, completed tasks = 4620]]"], [[xx-rr-1][86] - org.opensearch.OpenSearchException - "rejected execution of org.opensearch.common.util.concurrent.TimedRunnable@45d9c393 on EWMATrackingOpenSearchThreadPoolExecutor[name = 562a078f676b91f136fdd28bf0f75be7/write, queue capacity = 10000, task execution EWMA = 401.1ms, org.opensearch.common.util.concurrent.EWMATrackingOpenSearchThreadPoolExecutor@44415adc[Running, pool size = 32, active threads = 32, queued tasks = 10011, completed tasks = 4620]]"], [[xx-rr-1][27] - org.opensearch.OpenSearchException - "rejected execution of org.opensearch.common.util.concurrent.TimedRunnable@121fc524 on EWMATrackingOpenSearchThreadPoolExecutor[name = 562a078f676b91f136fdd28bf0f75be7/write, queue capacity = 10000, task execution EWMA = 401.1ms, org.opensearch.common.util.concurrent.EWMATrackingOpenSearchThreadPoolExecutor@44415adc[Running, pool size = 32, active threads = 32, queued tasks = 10000, completed tasks = 4620]]"], [[xx-rr-1][58] - org.opensearch.OpenSearchException - "rejected execution of org.opensearch.common.util.concurrent.TimedRunnable@73f98d3e on EWMATrackingOpenSearchThreadPoolExecutor[name = 562a078f676b91f136fdd28bf0f75be7/write, queue capacity = 10000, task execution EWMA = 401.1ms, org.opensearch.common.util.concurrent.EWMATrackingOpenSearchThreadPoolExecutor@44415adc[Running, pool size = 32, active threads = 32, queued tasks = 10009, completed tasks = 4620]]"], [[xx-rr-1][15] - org.opensearch.OpenSearchException - "rejected execution of org.opensearch.common.util.concurrent.TimedRunnable@5d4e3761 on EWMATrackingOpenSearchThreadPoolExecutor[name = 562a078f676b91f136fdd28bf0f75be7/write, queue capacity = 10000, task execution EWMA = 401.1ms, org.opensearch.common.util.concurrent.EWMATrackingOpenSearchThreadPoolExecutor@44415adc[Running, pool size = 32, active threads = 32, queued tasks = 10000, completed tasks = 4620]]"], [[xx-rr-1][14] - org.opensearch.OpenSearchException - "rejected execution of org.opensearch.common.util.concurrent.TimedRunnable@312f038e on EWMATrackingOpenSearchThreadPoolExecutor[name = 562a078f676b91f136fdd28bf0f75be7/write, queue capacity = 10000, task execution EWMA = 401.1ms, org.opensearch.common.util.concurrent.EWMATrackingOpenSearchThreadPoolExecutor@44415adc[Running, pool size = 32, active threads = 32, queued tasks = 10000, completed tasks = 4620]]"], [[xx-rr-1][46] - org.opensearch.OpenSearchException - "rejected execution of org.opensearch.common.util.concurrent.TimedRunnable@ed8c638 on EWMATrackingOpenSearchThreadPoolExecutor[name = 562a078f676b91f136fdd28bf0f75be7/write, queue capacity = 10000, task execution EWMA = 401.1ms, org.opensearch.common.util.concurrent.EWMATrackingOpenSearchThreadPoolExecutor@44415adc[Running, pool size = 32, active threads = 32, queued tasks = 10001, completed tasks = 4620]]"], [[xx-rr-1][76] - org.opensearch.OpenSearchException - "rejected execution of org.opensearch.common.util.concurrent.TimedRunnable@4c9f9430 on EWMATrackingOpenSearchThreadPoolExecutor[name = 562a078f676b91f136fdd28bf0f75be7/write, queue capacity = 10000, task execution EWMA = 401.1ms, org.opensearch.common.util.concurrent.EWMATrackingOpenSearchThreadPoolExecutor@44415adc[Running, pool size = 32, active threads = 32, queued tasks = 10009, completed tasks = 4620]]"], [[xx-rr-1][82] - org.opensearch.OpenSearchException - "rejected execution of org.opensearch.common.util.concurrent.TimedRunnable@470736e2 on EWMATrackingOpenSearchThreadPoolExecutor[name = 562a078f676b91f136fdd28bf0f75be7/write, queue capacity = 10000, task execution EWMA = 401.1ms, org.opensearch.common.util.concurrent.EWMATrackingOpenSearchThreadPoolExecutor@44415adc[Running, pool size = 32, active threads = 32, queued tasks = 10017, completed tasks = 4620]]"], [[xx-1][7] - org.opensearch.OpenSearchException - "rejected execution of org.opensearch.common.util.concurrent.TimedRunnable@74f7db9d on EWMATrackingOpenSearchThreadPoolExecutor[name = 562a078f676b91f136fdd28bf0f75be7/write, queue capacity = 10000, task execution EWMA = 401.1ms, org.opensearch.common.util.concurrent.EWMATrackingOpenSearchThreadPoolExecutor@44415adc[Running, pool size = 32, active threads = 32, queued tasks = 10002, completed tasks = 4620]]"], [[xx-rr-1][38] - org.opensearch.OpenSearchException - "rejected execution of org.opensearch.common.util.concurrent.TimedRunnable@65cc9307 on EWMATrackingOpenSearchThreadPoolExecutor[name = 562a078f676b91f136fdd28bf0f75be7/write, queue capacity = 10000, task execution EWMA = 401.1ms, org.opensearch.common.util.concurrent.EWMATrackingOpenSearchThreadPoolExecutor@44415adc[Running, pool size = 32, active threads = 32, queued tasks = 10000, completed tasks = 4620]]"], [[xx-rr-1][11] - org.opensearch.OpenSearchException - "rejected execution of org.opensearch.common.util.concurrent.TimedRunnable@1924a8ae on EWMATrackingOpenSearchThreadPoolExecutor[name = 562a078f676b91f136fdd28bf0f75be7/write, queue capacity = 10000, task execution EWMA = 401.1ms, org.opensearch.common.util.concurrent.EWMATrackingOpenSearchThreadPoolExecutor@44415adc[Running, pool size = 32, active threads = 32, queued tasks = 10004, completed tasks = 4620]]"], [[xx-rr-1][75] - org.opensearch.OpenSearchException - "rejected execution of org.opensearch.common.util.concurrent.TimedRunnable@13110bfa on EWMATrackingOpenSearchThreadPoolExecutor[name = 562a078f676b91f136fdd28bf0f75be7/write, queue capacity = 10000, task execution EWMA = 401.1ms, org.opensearch.common.util.concurrent.EWMATrackingOpenSearchThreadPoolExecutor@44415adc[Running, pool size = 32, active threads = 32, queued tasks = 10000, completed tasks = 4620]]"], [[sxx-rr-1][72] - org.opensearch.OpenSearchException - "rejected execution of org.opensearch.common.util.concurrent.TimedRunnable@ecacfe0 on EWMATrackingOpenSearchThreadPoolExecutor[name = 562a078f676b91f136fdd28bf0f75be7/write, queue capacity = 10000, task execution EWMA = 401.1ms, org.opensearch.common.util.concurrent.EWMATrackingOpenSearchThreadPoolExecutor@44415adc[Running, pool size = 32, active threads = 32, queued tasks = 10012, completed tasks = 4620]]"], [[xx-rr-1][61] - org.opensearch.OpenSearchException - "rejected execution of org.opensearch.common.util.concurrent.TimedRunnable@4f5cec55 on EWMATrackingOpenSearchThreadPoolExecutor[name = 562a078f676b91f136fdd28bf0f75be7/write, queue capacity = 10000, task execution EWMA = 401.1ms, org.opensearch.common.util.concurrent.EWMATrackingOpenSearchThreadPoolExecutor@44415adc[Running, pool size = 32, active threads = 32, queued tasks = 10000, completed tasks = 4620]]"], [[xx-rr-1][3] - org.opensearch.OpenSearchException - "rejected execution of org.opensearch.common.util.concurrent.TimedRunnable@4a7e6d26 on EWMATrackingOpenSearchThreadPoolExecutor[name = 562a078f676b91f136fdd28bf0f75be7/write, queue capacity = 10000, task execution EWMA = 401.1ms, org.opensearch.common.util.concurrent.EWMATrackingOpenSearchThreadPoolExecutor@44415adc[Running, pool size = 32, active threads = 32, queued tasks = 10000, completed tasks = 4620]]"], [[xx-rr-1][66] - org.opensearch.OpenSearchException - "rejected execution of org.opensearch.common.util.concurrent.TimedRunnable@5777b9a9 on EWMATrackingOpenSearchThreadPoolExecutor[name = 562a078f676b91f136fdd28bf0f75be7/write, queue capacity = 10000, task execution EWMA = 401.1ms, org.opensearch.common.util.concurrent.EWMATrackingOpenSearchThreadPoolExecutor@44415adc[Running, pool size = 32, active threads = 32, queued tasks = 10002, completed tasks = 4620]]"], [[xx-rr-1][33] - org.opensearch.OpenSearchException - "rejected execution of org.opensearch.common.util.concurrent.TimedRunnable@7765cd1d on EWMATrackingOpenSearchThreadPoolExecutor[name = 562a078f676b91f136fdd28bf0f75be7/write, queue capacity = 10000, task execution EWMA = 401.1ms, org.opensearch.common.util.concurrent.EWMATrackingOpenSearchThreadPoolExecutor@44415adc[Running, pool size = 32, active threads = 32, queued tasks = 10011, completed tasks = 4620]]"], [[xx-rr-1][32] - org.opensearch.OpenSearchException - "rejected execution of org.opensearch.common.util.concurrent.TimedRunnable@4d7cb3b5 on EWMATrackingOpenSearchThreadPoolExecutor[name = 562a078f676b91f136fdd28bf0f75be7/write, queue capacity = 10000, task execution EWMA = 401.1ms, org.opensearch.common.util.concurrent.EWMATrackingOpenSearchThreadPoolExecutor@44415adc[Running, pool size = 32, active threads = 32, queued tasks = 10000, completed tasks = 4620]]"], """,
  "leader_alias" : "xx-prod",
  "leader_index" : "xx",
  "follower_index" : "xx-1"
}
```

**What is your host/environment?**
 - OS: AWS Opensearch
 - Version 1.3
 - Plugins

### PR DESCRIPTION
**Pull request by sricharanvuppu**
Signed-off-by: sricharanvuppu <srivuppu@amazon.com>

### Description
Handles Opensearch reject execuition excpetion handling 
 
### Issues Resolved
#627 
 
### Testing 
Tested exception handling on live cluster 

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/kotlin/org/opensearch/replication/ReplicationException.kt [hunk 1]
```diff
@@ -12,22 +12,27 @@
...
  * [org.opensearch.index.shard.ShardNotFoundException].
  */
-class ReplicationException: OpenSearchException {
+class ReplicationException: OpenSearchStatusException {
 
-    constructor(message: String, vararg args: Any) : super(message, *args)
+    constructor(message: String, status : RestStatus, cause: Throwable,  vararg args: Any) : super(message, status, cause, *args)
 
-    constructor(message: String, cause: Throwable, vararg args: Any) : super(message,  cause, *args)
+    constructor(message: String, vararg args: Any) : super(message, RestStatus.INTERNAL_SERVER_ERROR, *args)
```
Comment (saikaranam-amazon) [inline src/main/kotlin/org/opensearch/replication/ReplicationException.kt:30]: Any reason to move to status exception?

File: src/main/kotlin/org/opensearch/replication/util/Extensions.kt [hunk 2]
```diff
@@ -130,6 +132,16 @@ suspend fun <Req: ActionRequest, Resp: ActionResponse> Client.suspendExecuteWith
                 throw e
             }
         }
+        catch (e: OpenSearchRejectedExecutionException) {
+            if(index < numberOfRetries-2) {
+                log.warn("Encountered a failure while executing in $req. Retrying in ${currentBackoff/1000} seconds" + ".", e)
+                delay(currentBackoff)
+                currentBackoff = (currentBackoff * factor).toLong().coerceAtMost(maxTimeOut)
+            }
```
Comment (saikaranam-amazon) [inline src/main/kotlin/org/opensearch/replication/util/Extensions.kt:140]: Could we move the common logic b/w the exceptions after the catch block?

File: src/main/kotlin/org/opensearch/replication/task/shard/TranslogSequencer.kt [hunk 3]
```diff
@@ -66,35 +72,57 @@ class TranslogSequencer(scope: CoroutineScope, private val replicationMetadata:
...
+                                    TimeUnit.NANOSECONDS.toMillis(tookInNanos)
+                                )
+                                followerClusterStats.stats[followerShardId]!!.opsWritten.addAndGet(
+                                    replayRequest.changes.size.toLong()
+                                )
+                            } catch (e: OpenSearchException) {
+                                if (e.status() == RestStatus.TOO_MANY_REQUESTS) {
+                                    tooManyRequestsAcknowledgement = true
+                                } else {
+                                    throw e
```
Comment (saikaranam-amazon) [inline src/main/kotlin/org/opensearch/replication/task/shard/TranslogSequencer.kt:118]: Once this exception is thrown, we will need to retry this at the task level.
Could we please check this case as well?

File: src/main/kotlin/org/opensearch/replication/ReplicationException.kt (hunk 1 above)
Comment (sricharanvuppu) [reply src/main/kotlin/org/opensearch/replication/ReplicationException.kt:30]: Replication exception throwing only INTERNAL_SERVER_ERROR exception and to enable throwing TOO_MANY_REQUESTS exception changed it to status exception

**Pull request closed without merging**